File: protocol-generator/php/src/Base58.php
<?php

declare(strict_types=1);

namespace EntityCore;

/**
 * Base58 (Bitcoin alphabet) encode/decode, hand-rolled (no Composer dep — dodges
 * a dep + a pin). Used for peer-id formatting/parsing (V7 §1.5). Each leading
 * zero byte maps to a leading "1", per the standard Base58 convention.
 *
 * Base58 needs unsigned arithmetic over arbitrary-size byte strings; PHP has no
 * native bignum, so this reuses the SAME GMP carrier as the codec
 * ([idiom].uint64_carrier) — no native-int path that could overflow to a lossy
 * float.
 */
final class Base58
{
    private const ALPHABET = '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz';

    /** Encode a binary string to a Base58 (ASCII) string. */
    public static function encode(string $bin): string
    {
        $len = \strlen($bin);
        $zeros = 0;
        while ($zeros < $len && $bin[$zeros] === "\x00") {
            $zeros++;
        }

        // GMP imports the bytes as a big-endian unsigned integer (no float path).
        $n = $len === 0 ? \gmp_init(0) : \gmp_import($bin, 1, \GMP_MSW_FIRST | \GMP_BIG_ENDIAN);

        $body = '';
        $fifty8 = \gmp_init(58);
        while (\gmp_cmp($n, 0) > 0) {
            [$n, $rem] = \gmp_div_qr($n, $fifty8);
            $body = self::ALPHABET[\gmp_intval($rem)] . $body;
        }

        return \str_repeat('1', $zeros) . $body;
    }

    /** Decode a Base58 string back to a binary string. */
    public static function decode(string $str): string
    {
        $len = \strlen($str);
        $ones = 0;
        while ($ones < $len && $str[$ones] === '1') {
            $ones++;
        }

        $n = \gmp_init(0);
        $fifty8 = \gmp_init(58);
        for ($i = 0; $i < $len; $i++) {
            $idx = \strpos(self::ALPHABET, $str[$i]);
            if ($idx === false) {
                throw new CodecException('invalid base58 character: ' . $str[$i]);
            }
            $n = \gmp_add(\gmp_mul($n, $fifty8), $idx);
        }

        $body = \gmp_cmp($n, 0) === 0 ? '' : \gmp_export($n, 1, \GMP_MSW_FIRST | \GMP_BIG_ENDIAN);

        return \str_repeat("\x00", $ones) . $body;
    }
}

File: protocol-generator/julia/src/base58.jl
# Base58 (Bitcoin alphabet) — hand-rolled (profile [codec].base58_library = hand-rolled).
# BigInt (GMP) makes base-256 ↔ base-58 long division trivial, so this is pure arithmetic,
# no byte-array carry loop. Leading zero bytes map to leading '1' characters (and back).
module Base58

export base58encode, base58decode

const ALPHABET = collect("123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz")

# reverse map: char → 0..57, or -1
const INV = let m = fill(-1, 128)
    for (i, c) in enumerate(ALPHABET)
        m[Int(c)+1] = i - 1
    end
    m
end

"""Base58-encode bytes → String. Leading 0x00 bytes become leading '1's."""
function base58encode(input::AbstractVector{UInt8})::String
    zeros = 0
    while zeros < length(input) && input[zeros+1] == 0x00
        zeros += 1
    end
    num = BigInt(0)
    for b in input
        num = num * 256 + Int(b)
    end
    digits = Char[]
    while num > 0
        num, rem = divrem(num, 58)
        push!(digits, ALPHABET[Int(rem)+1])
    end
    reverse!(digits)
    return String(vcat(fill('1', zeros), digits))
end

"""Base58-decode a String → bytes. Leading '1's restore leading 0x00 bytes."""
function base58decode(s::AbstractString)::Vector{UInt8}
    ones = 0
    for c in s
        c == '1' ? (ones += 1) : break
    end
    num = BigInt(0)
    for c in s
        ci = Int(c)
        (ci < 0 || ci >= 128 || INV[ci+1] < 0) && error("invalid base58 character: $c")
        num = num * 58 + INV[ci+1]
    end
    body = UInt8[]
    while num > 0
        num, rem = divrem(num, 256)
        push!(body, UInt8(rem))
    end
    reverse!(body)
    return vcat(fill(0x00, ones), body)
end

end # module Base58
